text/0018-generalized-in.md

```diff
@@ -0,0 +1,185 @@
+
+* `principal.manager in Group::"A"`
+
+This is true if `principal.manager` is a member of `Group::"A"`.
+
+* `principal.manager in [Group::"A", Group::"B"]`
+
+This is true if `principal.manager` is a member of either group. Equivalent to `principal.manager in Group::"A" || principal.manager in Group::"B"`.
+
+This has been referred to as the "deep" semantics of `in` -- the `x in [A, B]` form doesn't just check if `x` is equal to `A` or `B`, but (recursively) if `x` is a member of `A` or `B` in the hierarchy.
+
+In contrast, set membership is inherently shallow. If `in` was naively adapted to support set membership, the above example would be ambiguous: did the policy author mean set membership (in which case they're intending `principal.manager == Group::"A" || principal.manager == Group::"B"`), or today's Cedar's deep hierarchy membership (with the semantics above)?
+
+This RFC proposes to combine these two semantics in a back-compatible way.
+Informally, `in` is "deep" for entity elements, and shallow for non-entity elements.
+More formally, the semantics of `a in B` is:
+* if `a` and `B` are both entity references, then `true` if `a` is either equal to `B`, or a descendant of `B` in the hierarchy, else `false`
+    * this is not changed from today's Cedar
+* if `B` is a Cedar set and `a` is not an entity reference, then it's ordinary shallow set membership, i.e., equivalent to `B.contains(a)`
+    * this case is a type error in today's Cedar
+* if `B` is a Cedar set and `a` is an entity reference, then WLOG let `B` be `[B1, B2, B3]`. The semantics of `a in [B1, B2, B3]` is `a op B1 || a op B2 || a op B3` where for each `Bx`, if `Bx` is an entity reference then `op` is `in`, otherwise `op` is `==`
+    * in the case where `B` only contains entity references, this behavior matches today's Cedar
+    * in the case where `B` contains other values, this would be a type error in today's Cedar
+* in all other cases, type error
+    * this is not changed from today's Cedar
+
+Here's one additional example (we'll call it Example 3):
+```
+resource.owner in ["foo", "bar", User::"Group"]
+```
+Under this proposal, this expression is `true` if `resource.owner` is the string `"foo"`, the string `"bar"`, the entity reference `User::"Group"`, or any entity reference which is `in` `User::"Group"`.
+
+Arguably, this is the "intuitive" semantics that users want for `in` -- it blends "deep" `in` for entity references while also providing shallow set-membership for sets.
+
+One final note -- this RFC does not propose any changes to what syntax is legal in the policy scope. And of course, it doesn't change the behavior of any syntax that's currently legal in the policy scope.
+
+## Drawbacks
+
+- The new semantics is more complicated than Cedar's current semantics for `in`. This could lead to additional confusion.
+    - Response: The current semantics also leads to confusion, which this proposal at least partially addresses.
```

Comment:
The current semantics leads to confusion that is easily remedied (once you fix the error message): You really wanted `.contains` but you used `in` instead. Oops. Change your expression then you are done.

Reply:
Good point.  And for those tracking, the error message fix is cedar-policy/cedar#177

Reply:
Updated the text here (and under Alternative 1) in light of cedar#177